### PR TITLE
fix(web): make launcher shutdown cross-platform

### PR DESCRIPTION
**Pull request by Maxsky5**
## Summary
- Make the `@openducktor/web` launcher host force-stop path platform-aware.
- Preserve Unix/macOS process-group cleanup while avoiding negative-PID signaling on Windows.
- Add unit coverage for Unix, Windows, invalid PID, and graceful-exit shutdown behavior.

## Goal
`bunx @openducktor/web` starts a detached Rust web host and is responsible for shutting it down when the launcher exits or receives Ctrl+C. The previous force cleanup always used negative-PID process-group signals such as `process.kill(-pid, "SIGTERM")`, which are Unix-specific and unsafe for Windows.

This PR fixes that shutdown path without changing the frontend/backend protocol. The launcher still asks the host to shut down through `POST /shutdown` with `x-openducktor-control-token`, closes the frontend server, waits for the host to exit gracefully, and only force-stops the process if that wait times out.

## Technical choices
- Kept the process-management logic local to `packages/openducktor-web/src/launcher.ts` to avoid introducing a broader abstraction for a single launcher-specific need.
- Added a deterministic `terminateHostProcess` helper with injectable `platform`, `killProcess`, and `sleep` dependencies so platform behavior can be tested on any OS.
- Preserved the existing non-Windows behavior: negative-PID `SIGTERM`, child kill, bounded wait, process-group liveness probe, negative-PID `SIGKILL`, final child kill, and bounded wait.
- Added a Windows branch that uses only Bun child-process termination and bounded waits, never negative PID signaling.
- Added `waitForGracefulHostExit` so the force-stop helper is only called when the host does not exit during the graceful shutdown window.

## Verification
- `bun run lint`
- `bun run typecheck`
- `bun run test`
- `bun run build`
- `bun run check:rust`
- `bun run test:rust`
- `cargo fmt --all --check` from `apps/desktop/src-tauri`
- `cargo clippy --workspace --all-targets -- -D warnings` from `apps/desktop/src-tauri`
- `cargo build` from `apps/desktop/src-tauri`

## Summary by CodeRabbit

* **Bug Fixes**
  * More deterministic, cross-platform host shutdown with staged graceful wait before escalation to stronger termination methods.
  * Improved handling for Windows vs. non-Windows termination flows and edge cases where process groups are absent.

* **Tests**
  * Expanded tests covering graceful-exit outcomes, escalation timing, and platform-specific termination behaviors.

* **Chores**
  * No public API changes.

[Review Change Stack](https://app.coderabbit.ai/change-stack/Maxsky5/openducktor/pull/565)

### CODE REVIEW

File: packages/openducktor-web/src/launcher.ts [hunk 1]
```diff
@@ -141,32 +146,70 @@ const verifyBackendReadiness = async (
...
   const pid = child.pid;
-  if (typeof pid === "number" && pid > 0) {
+  const hasProcessGroupPid = typeof pid === "number" && pid > 0;
+
+  if (dependencies.platform === "win32") {
+    child.kill();
+    await waitForHostExit(child, 3_000, dependencies.sleep);
+
+    child.kill(9);
+    await waitForHostExit(child, 1_000, dependencies.sleep);
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply packages/openducktor-web/src/launcher.ts]: Fixed in 720eb2776 by using a Windows process-tree termination primitive (`taskkill /PID <pid> /T /F`) for valid Windows PIDs instead of only terminating the direct Bun child. The Windows path still never uses negative PID signaling.

File: packages/openducktor-web/src/launcher.ts [hunk 2]
```diff
@@ -433,9 +478,11 @@ export const runLauncher = async (options: LauncherOptions): Promise<number> =>
...
       }
 
-      await Promise.race([hostProcess.exited, Bun.sleep(HOST_GRACEFUL_EXIT_TIMEOUT_MS)]);
-      logInfo("Ensuring OpenDucktor host process has exited...");
-      await terminateProcessGroup(hostProcess);
+      const hostExited = await waitForGracefulHostExit(hostProcess);
+      if (!hostExited) {
+        logInfo("Ensuring OpenDucktor host process has exited...");
+        await terminateHostProcess(hostProcess);
+      }
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply packages/openducktor-web/src/launcher.ts:499]: Leaving this behavior as-is intentionally: the approved shutdown contract says to force-stop only if needed. If `/shutdown` succeeds and the host exits within the graceful window, we treat that as successful host-owned cleanup rather than force-killing after a clean exit.

File: packages/openducktor-web/src/launcher.test.ts [hunk 3]
```diff
@@ -97,6 +111,82 @@ describe("launcher internals", () => {
...
+      sleep: async () => {},
+    });
+
+    expect(processKillCalls).toEqual([
+      { pid: -1234, signal: "SIGTERM" },
+      { pid: -1234, signal: 0 },
+      { pid: -1234, signal: "SIGKILL" },
+    ]);
+    expect(killCalls).toEqual([undefined, 9]);
+  });
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply packages/openducktor-web/src/launcher.test.ts:154]: Fixed in 720eb2776 by adding coverage for the Unix path where `kill(-pid, 0)` throws because the process group is already gone, asserting `SIGKILL` is skipped.

File: packages/openducktor-web/src/launcher.ts [hunk 4]
```diff
@@ -141,32 +146,70 @@ const verifyBackendReadiness = async (
...
+  const hasProcessGroupPid = typeof pid === "number" && pid > 0;
+
+  if (dependencies.platform === "win32") {
+    child.kill();
+    await waitForHostExit(child, 3_000, dependencies.sleep);
+
+    child.kill(9);
+    await waitForHostExit(child, 1_000, dependencies.sleep);
+    return;
+  }
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply packages/openducktor-web/src/launcher.ts:205]: Fixed in 720eb2776 as part of the Windows tree-termination path: valid Windows PIDs now use the process-tree terminator instead of blindly calling `child.kill(9)` after the wait.

File: packages/openducktor-web/src/launcher.ts [hunk 5]
```diff
@@ -141,32 +146,70 @@ const verifyBackendReadiness = async (
...
       try {
-        process.kill(-pid, "SIGKILL");
+        dependencies.killProcess(-pid, "SIGKILL");
       } catch {}
     } catch {}
   }
 
   child.kill(9);
-  await Promise.race([child.exited, Bun.sleep(1_000)]);
+  await waitForHostExit(child, 1_000, dependencies.sleep);
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply packages/openducktor-web/src/launcher.ts]: Fixed in 720eb2776 by preserving the full Unix process-group grace sleep before probing the group and escalating to `SIGKILL`, plus a regression test where the direct child exits before the group is considered gone.
